### ISSUE TITLE
Unable to replicate continual learning experiment

### ISSUE THREAD

**Issue by rafaelcp** [ZiyaoLi/fast-kan#7]
I was trying to replicate the [continual learning experiment](https://github.com/KindXiaoming/pykan/blob/master/tutorials/Example_7_continual_learning.ipynb) from the KAN paper using fast-kan, without success. Upon inspection of values sent to the rbf after layernorm (fastkan.py:69), they were all zeros, besides the original x being a linspace from -1 to 1 with 500 steps. Thus, it wasn't able to learn anything. By removing the layernorm, I was able to replicate the original experiment. Since layernorm has normalized_shape=1, it treats each value as a single feature and as its own mean, thus, resulting in zeros.

Everything is the same as the original experiment as linked above, except for the training itself:

```
model = KAN([1, 1], use_base_update=False, num_grids=40, grid_min = -1, grid_max = 1) # use_base_update=True also works with SGD, but not with Adam (unless you set bias=False on line 65)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)
optimizer = optim.SGD(model.parameters(), lr=1) # Since the updates are very localized, we can use huge lr. Also, momentum and weight decay should be detrimental to continual learning.

# Define loss
criterion = nn.MSELoss()
ys = []

for group_id in range(n_peak):
    dataset = {}
    dataset['train_input'] = x_sample[group_id*n_num_per_peak:(group_id+1)*n_num_per_peak][:,None] # Tried full batches as well
    dataset['train_label'] = y_sample[group_id*n_num_per_peak:(group_id+1)*n_num_per_peak][:,None]
    dataset['test_input'] = x_sample[group_id*n_num_per_peak:(group_id+1)*n_num_per_peak][:,None]
    dataset['test_label'] = y_sample[group_id*n_num_per_peak:(group_id+1)*n_num_per_peak][:,None]
    for epoch in range(100):
      optimizer.zero_grad()
      output = model(dataset['train_input'].view(-1,1).to(device))
      loss = criterion(output, dataset['train_label'].to(device))
      loss.backward()
      optimizer.step()
    y_pred = model(x_grid[:,None].to(device))
    ys.append(y_pred.cpu().detach().numpy()[:,0])
```
Am I doing something wrong or that layernorm shouldn't indeed be there (at least for 1d data)?

**Comment by ZiyaoLi** [comment]
of course you cannot do layernorm on 1d data, because for 1d data x-x.mean() is always 0 so all outputs would be constants.

---- Replied Message ----
| From | Rafael ***@***.***> |
| Date | 05/24/2024 05:16 |
| To | ZiyaoLi/fast-kan ***@***.***> |
| Cc | Subscribed ***@***.***> |
| Subject | [ZiyaoLi/fast-kan] Unable to replicate continual learning experiment (Issue #7) |

I was trying to replicate the continual learning experiment from the KAN paper using fast-kan, without success. Upon inspection of values sent to the rbf after layernorm (fastkan.py:69), they were all zeros, besides the original x being a linspace from -1 to 1 with 500 steps. Thus, it wasn't able to learn anything. By removing the layernorm, I was able to replicate the original experiment. Since layernorm has normalized_shape=1, it treats each value as a single feature and as its own mean, thus, resulting in zeros.

Everything is the same as the original experiment as linked above, except for the training itself:

model = KAN([1, 1], use_base_update=False, num_grids=40, grid_min = -1, grid_max = 1) # use_base_update=True also works with SGD, but not with Adam (unless you set bias=False on line 65)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)
optimizer = optim.SGD(model.parameters(), lr=1) # Since the updates are very localized, we can use huge lr. Also, momentum and weight decay should be detrimental to continual learning.

# Define loss
criterion = nn.MSELoss()
ys = []

for group_id in range(n_peak):
    dataset = {}
    dataset['train_input'] = x_sample[group_id*n_num_per_peak:(group_id+1)*n_num_per_peak][:,None] # Tried full batches as well
    dataset['train_label'] = y_sample[group_id*n_num_per_peak:(group_id+1)*n_num_per_peak][:,None]
    dataset['test_input'] = x_sample[group_id*n_num_per_peak:(group_id+1)*n_num_per_peak][:,None]
    dataset['test_label'] = y_sample[group_id*n_num_per_peak:(group_id+1)*n_num_per_peak][:,None]
    for epoch in range(100):
      optimizer.zero_grad()
      output = model(dataset['train_input'].view(-1,1).to(device))
      loss = criterion(output, dataset['train_label'].to(device))
      loss.backward()
      optimizer.step()
    y_pred = model(x_grid[:,None].to(device))
    ys.append(y_pred.cpu().detach().numpy()[:,0])


Am I doing something wrong or that layernorm shouldn't indeed be there (at least for 1d data)?

**Comment by rafaelcp** [comment]
You're welcome!
